### PR TITLE
modified grpc_endpoint vtable and implementations

### PR DESCRIPTION
**Pull request by michaelywg**
This pull request adds a get_local_address() function to the grpc_endpoint API. The newly implemented grpc_endpoint_get_local_address() returns a std::string containing a URI representation of the local address for a grpc_endpoint connection.This functionality is analogous to the existing function grpc_endpoint_get_peer(), which returns the URI of the peer address for a grpc_endpoint socket. In addition to adding get_local_address() to the grpc_endpoint vtable and appropriately changing all existing endpoint implementations, this pull request changes the return type of grpc_endpoint_get_peer() to be std::string instead of char*.

### CODE REVIEW

**Comment by michaelywg** [comment]
Thank you for the review. I have made changes to address these issues.

File: test/core/util/reconnect_server.cc [hunk 1]
```diff
@@ -59,27 +60,25 @@ static void on_connect(void* arg, grpc_endpoint* tcp,
...
+  last_colon = peer.rfind(':');
+  if (server->peer == nullptr) {
+    server->peer = new std::string(peer);
+  } else {
+    if (last_colon == std::string::npos) {
+      gpr_log(GPR_ERROR, "peer does not contain a ':'");
+    } else if (peer.compare(0, static_cast<size_t>(last_colon),
+                            *server->peer) != 0) {
+      gpr_log(GPR_ERROR, "mismatched peer! %s vs %s", server->peer->c_str(),
+              peer);
```
Comment (ashithasantosh) [inline test/core/util/reconnect_server.cc:81]: We may need peer.data() here.

File: src/core/lib/iomgr/tcp_windows.cc [hunk 2]
```diff
@@ -514,17 +521,28 @@ grpc_endpoint* grpc_tcp_create(grpc_winsocket* socket,
...
+    grpc_resolved_address resolved_local_addr;
+    memcpy(resolved_local_addr.addr, &local_addr, local_addrlen);
+    resolved_local_addr.len = local_addrlen;
+    tcp->local_address = grpc_sockaddr_to_uri(&resolved_local_addr);
+  }
+  tcp->peer_string = peer_string;
   grpc_slice_buffer_init(&tcp->last_read_buffer);
-  tcp->resource_user = grpc_resource_user_create(resource_quota, peer_string);
+  tcp->resource_user =
+      grpc_resource_user_create(resource_quota, peer_string.c_str());
```
Comment (ashithasantosh) [inline src/core/lib/iomgr/tcp_windows.cc:545]: We don't need the conversion here. It can just be peer_string.

**Comment by michaelywg** [comment]
Thanks for the review. The merge conflict should be fixed now. I added a constructor for grpc_tcp in tcp_posix.cc to prevent `tcp_zerocopy_send_ctx` from being initialized twice (and thus leaking memory) when using `new` in `grpc_tcp_create`.

**Comment by michaelywg** [comment]
Thank you for the review. I have squashed all my changes to one commit.

**Pull request merged**